lms/djangoapps/instructor/views/api.py

```diff
@@ -509,8 +505,85 @@ def create_and_enroll_user(email, username, name, country, password, course_id):
+            un-enrolled from enrolled etc.
+    :return CourseEnrollment instance.
+    """
+    enrollment_obj = CourseEnrollment.enroll(enrolled_user, course_id, mode=mode)
+    ManualEnrollmentAudit.create_manual_enrollment_audit(
+        enrolled_by, enrolled_user.email, state_transition, reason, enrollment_obj
+    )
+
+    log.info(u'user %s enrolled in the course %s', enrolled_user.username, course_id)
+    return enrollment_obj
+
+
+def create_and_enroll_user(email, username, name, country, password, course_id, course_mode, enrolled_by, email_params):
+    """
+    Create a new user and enroll him/her to the given course, return list of errors in the following format
+        Error format:
+            each error is key-value pait dict with following key-value pairs.
+            1. username: username of the user to enroll
+            1. email: email of the user to enroll
+            1. response: readable error message
+
+    :param email: user's email address
+    :param username: user's username
+    :param name: user's name
+    :param country: user's country
+    :param password: user's password
+    :param course_id: course identifier of the course in which to enroll the user.
+    :param course_mode: mode for user enrollment, e.g. 'honor', 'audit' etc.
+    :param enrolled_by: User who made the manual enrollment entry (usually instructor or support)
+    :param email_params: information to send to the user via email
+
+    :return: list of errors
+    """
+    errors = list()
+    try:
+        with transaction.atomic():
+            # Create a new user
+            user = create_user_and_user_profile(email, username, name, country, password)
+
+            # Enroll user to the course and add manual enrollment audit trail
+            create_manual_course_enrollment(
+                enrolled_user=user,
+                course_id=course_id,
+                mode=course_mode,
+                enrolled_by=enrolled_by,
+                reason='Enrolling via csv upload',
+                state_transition=UNENROLLED_TO_ENROLLED,
+            )
+    except IntegrityError:
+        errors.append({
+            'username': username, 'email': email, 'response': _('Username {user} already exists.').format(user=username)
+        })
+    except Exception as ex:  # pylint: disable=broad-except
+        log.exception(type(ex).__name__)
+        errors.append({
+            'username': username, 'email': email, 'response': type(ex).__name__,
+        })
+    else:
+        # It's a new user, an email will be sent to each newly created user.
+        email_params['message'] = 'account_creation_and_enrollment'
+        email_params['email_address'] = email
+        email_params['password'] = password
+        email_params['platform_name'] = microsite.get_value('platform_name', settings.PLATFORM_NAME)
+        send_mail_to_student(email, email_params)
```

Comment:
I think my only concern with this 'else' block is that we're not just doing some follow-up work to the 'try', such as writing a log entry.  We're invoking a subflow on another component and are doing so in an unprotected manner.  If we would like to keep the code structured this way, we should think about wrapping send_mail_to_student in an additional try/except to protect against exceptions raised while working with this component.  

Otherwise this changeset looks really good to me!